src/azul/portal_service.py

```diff
@@ -0,0 +1,82 @@
+from copy import deepcopy
+from typing import (
+    Sequence,
+)
+import json
+import logging
+
+import boto3
+from botocore.exceptions import ClientError
+
+from azul import config
+from azul.plugin import Plugin
+from azul.types import JSON
+
+log = logging.getLogger(__name__)
+
+
+class PortalService:
+
+    def get_portal_integrations_db(self) -> Sequence[JSON]:
+        """
+        Retrieve portal DB from S3.
+
+        If no object is present at the location specified in azul.config, then
+        a default hardcoded DB is obtained from the current plugin, uploaded to
+        S3, and returned.
+        """
+
+        client = boto3.client('s3')
+        try:
+            retrieval = client.get_object(Bucket=config.portal_integrations_db_bucket,
+                                          Key=config.portal_integrations_db_object)
+        except client.exceptions.NoSuchKey:
+            plugin = Plugin.load()
+            db = self.stage_db(plugin.portal_integrations_db())
+            log.info('Portal integration DB not found in S3; using and uploading hardcoded DB from plugin.')
+            json_bytes = bytes(json.dumps(db), 'UTF-8')
+            client.put_object(Bucket=config.portal_integrations_db_bucket,
+                              Key=config.portal_integrations_db_object,
+                              Body=json_bytes,
+                              ContentType='application/json')
+        except ClientError as e:
+            log.error(f'Unexpected error retrieving portal DB from S3: {e}')
+            raise
+        else:
+            log.info("Retrieved portal integration DB from S3")
+            json_bytes = retrieval['Body'].read()
+            db = json.loads(json_bytes.decode('UTF-8'))
+        return db
+
+    def stage_db(self, unstaged_db: Sequence[JSON]) -> Sequence[JSON]:
+        """
+        Transform portal integrations database to only contain entity_ids from
+        the current DSS deployment stage, leaving the original unmodified.
+
+        :param unstaged_db: portal DB where the `entity_ids` fields  are
+        dictionaries whose keys correspond to DSS deployment stages.
+
+        :return: deep copy of that DB where the `entity_ids` fields have been
+        replaced by the entry associated with the current DSS deployment stage.
+        If the `entity_ids` field is present but no entity ids are specified for
+        the current deployment stages, the integration is removed. Portals,
+        however, are not removed even if they have no remaining associated
+        integrations.
+        """
+
+        db = deepcopy(unstaged_db)
```

Comment:
```suggestion
        db = deepcopy(db)
```

I like this because it signals that you are done with the original argument value from this point forward. It also means that you don't need to invent a discriminating prefix or suffix, like `staged` or `unstaged` in order to distinguish between variable.